File: src/main/kotlin/com/vlnabatov/alabaster/extensions/annotation/GoAnnotator.kt
package com.vlnabatov.alabaster.extensions.annotation

import com.goide.GoParserDefinition.Lazy.STRING_LITERALS
import com.goide.GoTypes.IDENTIFIER
import com.intellij.lang.annotation.AnnotationHolder
import com.intellij.lang.annotation.Annotator
import com.intellij.lang.annotation.HighlightSeverity.INFORMATION
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors.BRACES
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors.NUMBER
import com.intellij.psi.PsiElement
import com.intellij.psi.impl.source.tree.LeafPsiElement
import com.vlnabatov.alabaster.annotateSeparationMarks


class GoAnnotator : Annotator {
    override fun annotate(element: PsiElement, holder: AnnotationHolder) {
        if (element !is LeafPsiElement) return

        try {
            when (element.elementType) {
                in STRING_LITERALS -> annotateSeparationMarks(element, holder, BRACES)
                IDENTIFIER if (element.text == "nil") ->
                    holder.newSilentAnnotation(INFORMATION).textAttributes(NUMBER).create()
            }
        } catch (e: Exception) {
            /* Should not happen */
        }


    }
}

File: src/main/clojure/com/vlnabatov/alabaster/go_annotator.clj
(ns com.vlnabatov.alabaster.go-annotator
  (:require [com.vlnabatov.alabaster.util :refer [import-static print-methods]]
            [clojure.pprint :as pprint]
            [clojure.reflect :as reflect]
            [clojure.inspector :as insp])
  (:import (com.goide.psi GoTokenType)
           (com.intellij.lang.annotation AnnotationHolder)
           (com.intellij.lang.annotation HighlightSeverity)
           (com.intellij.openapi.editor DefaultLanguageHighlighterColors)
           (com.goide GoParserDefinition$Lazy)
           (com.goide GoTypes)
           (com.intellij.psi PsiElement)
           (com.intellij.psi.tree IElementType)
           (com.vlnabatov.alabaster UtilKt))
  (:gen-class :implements [com.intellij.lang.annotation.Annotator]
              :name       com.vlnabatov.alabaster.extensions.annotation.GoAnnotator
              :prefix     "-"))



(defn -annotate
  [_ ^PsiElement element ^AnnotationHolder holder]
  (let [^IElementType elementType (.. element getNode getElementType)]
    (cond (.contains GoParserDefinition$Lazy/STRING_LITERALS elementType) (do (UtilKt/annotateSeparationMarks element
                                                                                                              holder))
          (and (= "nil" (.getText element)) (= elementType GoTypes/IDENTIFIER))
            (do (-> holder
                    (.newSilentAnnotation HighlightSeverity/INFORMATION)
                    (.textAttributes DefaultLanguageHighlighterColors/NUMBER)
                    (.create))))))
